### PR TITLE
Skip pub structs with repr(c) and repr(transparent) in dead code analysis

### LINKED ISSUE #126169: dead_code lint wrongly warns about public repr(C) structs with private fields but no constructors
### Code

```Rust
#[repr(C)] // <--
pub struct Foo {
    pub i: i16,
    align: i16
}
```

### Current output

```Shell
warning: field `align` is never read
 --> src/lib.rs:3:5
  |
1 | pub struct Foo {
  |            --- field in this struct
2 |     pub i: i16,
3 |     align: i16
  |     ^^^^^
  |
  = note: `#[warn(dead_code)]` on by default
```

### Desired output

```Shell
No warnings, or perhaps a different warning.
```

### Rationale and extra context

This warning is probably triggering because there's no constructor that allows a user to set the `align` field on struct `Foo`.  However, there is still a valid way for a user to construct it: `std::mem::zeroed()`.  The libc crate contains many structs that must be initialized that way.  libc could silence these warnings by `#[allow(dead_code)]`, but that's too broad a brush.  It would silence many much more serious warnings.  Could we please either revert to the previous behavior, or else create a new lint for this that can be separately silenced?

### Other cases

_No response_

### Rust Version

```Shell
rustc 1.80.0-nightly (804421dff 2024-06-07)
binary: rustc
commit-hash: 804421dff5542c9c7da5c60257b5dbc849719505
commit-date: 2024-06-07
host: x86_64-unknown-freebsd
release: 1.80.0-nightly
LLVM version: 18.1.7
```

### Anything else?

This is a regression since `rustc 1.80.0-nightly (f67a1acc0 2024-06-01)`.  That version did not warn in such cases.

Downstream issue: https://github.com/rust-lang/libc/issues/3740

### PR DESCRIPTION
**Pull request by mu001999**
Fixes #126169

### CODE REVIEW

File: compiler/rustc_passes/src/dead.rs [hunk 1]
```diff
@@ -931,7 +931,10 @@ fn create_and_seed_worklist(
                     match tcx.def_kind(id) {
                         DefKind::Impl { .. } => false,
                         DefKind::AssocConst | DefKind::AssocTy | DefKind::AssocFn => !matches!(tcx.associated_item(id).container, AssocItemContainer::ImplContainer),
-                        DefKind::Struct => struct_all_fields_are_public(tcx, id.to_def_id()) || has_allow_dead_code_or_lang_attr(tcx, id).is_some(),
+                        DefKind::Struct => struct_all_fields_are_public(tcx, id.to_def_id()) || has_allow_dead_code_or_lang_attr(tcx, id).is_some() || {
+                            let def = tcx.adt_def(id);
+                            def.repr().c() || def.repr().transparent()
```
Comment (workingjubilee) [inline compiler/rustc_passes/src/dead.rs:936]:
```suggestion
                            // often declared in Rust but constructed by FFI, so ignore
                            def.repr().c() || def.repr().transparent()
```

**Comment by compiler-errors** [comment]
I think any changes to lints should go through T-lang. Sorry for taking long to review this.

**Comment by tgross35** [comment]
This came up on Zulip https://rust-lang.zulipchat.com/#narrow/stream/122651-general/topic/Status.20of.20fix.20for.20dead_code.20warnings.20on.20repr.28C.29.20structs. It seems rather unfortunate that large bindings crates are needing to `allow(dead_code)` globally in order to work around this ([`libc`](https://github.com/rust-lang/libc/pull/3749), [`windows`](https://github.com/microsoft/windows-rs/pull/3098)).

If this proposed solution is decided not to be acceptable, unfortunately we should probably consider temporarily reverting or adjusting part of https://github.com/rust-lang/rust/pull/125572. One suggestion is to emit a new lint within `dead_code` that can be disabled separately, rather than `allow(dead_code)` being the only way to opt out https://github.com/rust-lang/rust/issues/126169#issuecomment-2166071230

Unfortunately this behavior just branched, so:

@rustbot label +beta-nominated

**Comment by workingjubilee** [comment]
This PR has been introduced to address a diagnostics regression. The original PR that introduced the behavior this change was meant to prevent regressing can no longer be cleanly reverted on beta. This is thus now the most minimal diff if we want to apply the change to beta, even if it means we introduce new code.

**Comment by compiler-errors** [comment]
See https://github.com/rust-lang/rust/pull/128329#issuecomment-2258477039

I think that all of this new dead code analysis landed recently should probably be reverted and reworked, since then we're not pressured to put up hacky fixes in favor of getting it correct and well documented the first time 👍

**Comment by Mark-Simulacrum** [comment]
Dropping beta-nominated, looks like the original PR was reverted on beta 1.81 (via https://github.com/rust-lang/rust/pull/128606).

**Pull request closed without merging**